### ISSUE TITLE
Optimise fast connections over slow connections for torrents with many seeds

### ISSUE THREAD

**Issue by Sophist-UK** [arvidn/libtorrent#4989]
**Please provide the following information**

libtorrent version (or branch): 1.2.6.0 (in latest version of qBitTorrent)

platform/architecture: Windows 10 64-bit

compiler and compiler version: Not known

When downloading a torrent with a large number of seeds, I get some connections which download at several 100Kbps, and some that download at a few bytes per second - and I would like LibTorrent to drop these in the hope that it can find a significantly faster seed. I also get a lot of connections to seeds which wait some time before downloading. I would like LibTorrent to:

a. For a torrent with many more seeds than allowed connections (setting to define the multiple), drop any connection which downloads at a speed much slower than the average of the fastest 50% of all other connections (setting to define the %). So if I have 3 connections at > 100Kbps and two which are below 1Kbps then the average speed of the top 2 would be c. 100Kbs; the bottom one is < 5% of this (i.e. below 5Kbps) so it gets dropped.

b. Ideally I would like the BitTorrent protocol to allow seed to estimate the time before it starts downloading and to estimate the maximum upload speed, so that LibTorrent can decide whether to hold on to that connection or try to find another connection which will start faster and / or download faster (using the above algorithm for the latter - just without waiting for the download to actually start and stabilise).  But I appreciate that this might need enhancements to the BitTorrent protocol and that this might therefore need to be a long-term aspiration - in which case I would like LibTorrent to do the best it can to optimise connections which actually download over those which wait a long time (e.g. by caching the wait time and speed for each IP address).

Obviously, for torrents with a limited number of seeds, a slow download is better than none.

**Comment by Sophist-UK** [comment]
To make it easier to understand, here is a screen shot. As you can see I have 4 seeds downloading at over 100KB/s, and 3 downloading at under 10KB/s. I would like LibTorrent to cancel the 3 slow ones and try to find 3 much faster ones instead.

![image](https://user-images.githubusercontent.com/3001893/90282161-2cf14680-de6e-11ea-9d85-db805df3c1a8.png)

**Comment by arvidn** [comment]
I also notice that you only have 15 peers. The default limit is 50 peers per torrent. So I would expect libtorrent to *already* be trying to connect more peers, regardless of whether the existing slow ones are left connected or not.

Do you have a reason to believe that disconnecting the slow ones would improve your chances of finding faster seeds?

Apart from that, let's say you *did* have the full 50 peers; there is no obvious way to implement what you ask. I welcome specific ideas though. One really important aspect is that, even if a change *empirically* improves your situation, there's always a risk it degrades the majority of situations.

Here are some of the challenges:

1. You can't know your own download capacity. If you connect a peer, and you can only get 10 kiB/s from it. How can you know this is because *that* peer has reached its upload capacity, or because *you* have reached your download capacity?

The optimal action for you to take heavily depends on that. If it's the former you probably want to connect more peers, if it's the latter you probably want to disconnect some peers.

2. A slow peer now may become fast later (if demand for its bandwidth is reduced). You can't know that there is another peer that can give you better bandwidth now so pre-emptively disconnecting a peer in the hope of finding a better one is always associated with risk. To minimise that risk you want to disconnect the slowest peer that has been connected the longest (i.e. least useful peer). But if you're not at capacity yet, you don't want to disconnect peers at all. I think your suggestion of disconnecting half of the peers would be too risky. What if you can't find that many new peers in time for this logic to run again? Then you'll disconnect the next half slowest peers. For example, once you've seen all peers in the swarm, you'll just disconnect peers until you're down to 2 or 3 peers left. (until some timeout expires and you think it's OK to connect back to the peers you disconnected earlier).

Your bullet (b) is not practically implementable. Estimating how long it will take until a peer is unchoked *could* be done, if the client implemented its choker in a certain way. But how often it would be right is hard to say. It would depend on churn among peers. And estimating ones upload capacity *could* be done by storing the recent max upload rate. But this might be *wildly* wrong, if you've only been connected to "slow" peers recently, and nobody has been able to saturate your upload link yet. I think the utility of that would be questionable also, since the rate *you* would get from that peer depend on so many other things, like the number of peers that seed is uploading to, and network conditions along the paths to all those peers, and to you.

Lastly, there is logic that, sort of, does (a) already, but less aggressive. See [peer_turnover](https://www.libtorrent.org/reference-Settings.html#peer_turnover).

**Comment by Sophist-UK** [comment]
Many home routers have a limit on the concurrent TCP connections - hence I keep the number of connections down. However my setting is 25, so you are correct that this screenshot is below the limit of connections. I cannot recall whether at the time of this screenshot there were other torrents which were using connections, but that is likely.

1. Most people do know their download capacity. In my case it is 40Mb/s which is my internet connection speed and which equates to 5MB/s. I have seen torrent downloads up to 2MB/s (but have no reason to believe it could not go higher if I had enough high bandwidth connections), and I have seen other software achieve above 4MB/s. But I accept that LibTorrent cannot know the download bandwidth - but the proposed algorithm only needs to compare the download speed of one connection to that of other connections and to know that there are many other seeds that it could potentially connect to. In the circumstances where connection 1 is downloading at (say) <5% of connection 2, we can reasonably assume that connection 1 is limited by upload bandwidth (or possibly an intermediate link) because if this wasn't the case and it was download bandwidth limited then both connections would likely be download at similar speeds.

2A. Yes - it is possible that later a connection which is currently slow may get much faster. However... A) do we want to wait on the off chance that it will speed up, or would it be better to cut our losses and find a faster peer? And B) we are not risking much by disconnecting a peer which is <5% of other seeds (because the other much faster seeds will take <1/20 of the time to download what we decide to forego, but we stand to gain a lot by finding a much faster peer.

2B. Yes - I agree we should not disconnect peers when we have available connections both globally and on this particular torrent, because we don't need to lose the slow download to get the benefits of another connection.

2C. I agree that we need to avoid an algorithm which avoids connecting to peers and doesn't try to connect as many peers as it is allowed to. This algorithm should only disconnect one very slow peer when connections (global or torrent) are maximum and there are a lot of other peers to try and it stands to make a big gain in speed from another peer.

2D. The record of wait time before downloading started (W), downloading period (D), download volume (V) for this connection and last-connected / connection attempt time (T) should be used to order the list of peers to try by calculating the average speed S = V/(W+D) - perhaps using the following algorithm: 

i) IP addresses without measurements or with T older than "IP download maximum cache time" are ordered first (so we attempt to connect to all peers);
ii) then those peers with T between "IP download maximum cache time" and "IP download minimum cache time" (setting) and with previous measurements are in decreasing order by speed S=V/(W+D);
iii) IP addresses with T newer than "IP download minimum cache time limit" are ordered last by S descending.

3. peer_turnover - yes - it does appear that peer_turnover is an unintelligent mechanism to avoid LibTorrent staying connected to the same peers and not finding faster ones - but it is completely unintelligent, selecting connections at random (or the oldest ones) to disconnect rather than choosing slow ones where there is a high probability of finding faster ones. I guess that one implementation approach would be to still use the peer_turnover mechanism to disconnect a proportion of connections on a periodic basis rather than when the number of free connection slots (globally or for a torrent) gets to zero, just make a more intelligent choice. Alternatively you could still keep a periodic random disconnection (to ensure that you don't hog a connection to a peer) and have a parallel implementation of disconnecting slow connections when connection slots are exhausted.

Note: qBitTorrent does not (AFAIK) currently provide Advanced Settings for peer_turnover. I will add an enhancement request to add these.

**Comment by arvidn** [comment]
>  And B) we are not risking much by disconnecting a peer which is <5% of other seeds (because the other much faster seeds will take <1/20 of the time to download what we decide to forego, but we stand to gain a lot by finding a much faster peer.

That's not really true. If all your peers are sending data to you at essentially the same rate, the bottom 5th percentile of peers will be essentially some random ones (because of noise). You can't assume that the download rate distribution always looks the way you see it. It's actually quite common that, when you saturate your download link, almost all peers are sending to you at the same rate (because everyone is subject to *your* capacity limit).

**Comment by arvidn** [comment]
> selecting connections at random (or the oldest ones) to disconnect rather than choosing slow ones where there is a high probability of finding faster ones.

Why do you say peers are disconnected at random?

> just make a more intelligent choice.

Can you be more specific?

**Comment by arvidn** [comment]
here's the comparison function used to pick peers to disconnect:

https://github.com/arvidn/libtorrent/blob/RC_1_2/src/torrent.cpp#L7414

It looks like there are probably a few improvements that could be made to it. Do you want to propose any?

**Comment by Sophist-UK** [comment]
> That's not really true. If all your peers are sending data to you at essentially the same rate, the bottom 5th percentile of peers will be essentially some random ones (because of noise). You can't assume that the download rate distribution always looks the way you see it. It's actually quite common that, when you saturate your download link, almost all peers are sending to you at the same rate (because everyone is subject to _your_ capacity limit).

**Exactly!!**

So if you have two peers one downloading at > 100KBps and one <1KBps then you can be pretty certain that the 1Kbps connection is NOT limited by your own bandwidth and is therefore limited by the sender. If it wasn't, as you say, they would both have approximately the same download speed. (Note: This does not mean that the faster one isn't being limited by your own bandwidth limit, just that the slow one is limited by the sender.)

> here's the comparison function used to pick peers to disconnect:
>
> https://github.com/arvidn/libtorrent/blob/RC_1_2/src/torrent.cpp#L7414
>
> It looks like there are probably a few improvements that could be made to it. Do you want to propose any?

Thanks for this pointer. The comparison function is actually pretty good and certainly **not** random. TBH this is almost but not quite as good as doing it when connection slots are exhausted - I will see whether it is possible to run it when connection slots are exhausted as well as on a timer. I will need to look at the code in more detail, but I suspect that it will disconnect connections which have connected but not started downloading (download speed = 0) in preference to those that are downloading VERY slowly - in which case there is scope for tweaking. Also it is highly unlikely that the download speeds will be the same (unless both have downloaded nothing), so I suspect that there is scope for some improvements.

However, this does not influence making new connections to IP addresses which are known to be faster. That would be quite beneficial I think, but a bit more complex to implement.

I am not really a C / C+ coder, so I can attempt to tweak source code, but I cannot compile and test.

**Comment by arvidn** [comment]
This logic runs on a timer *AND* if the connection limit has been reached. Both global and per torrent in fact. The overarching (session-wide) trigger for this is here:

https://github.com/arvidn/libtorrent/blob/RC_1_2/src/session_impl.cpp#L3458

**Comment by Sophist-UK** [comment]
Any chance of a pointer to the code which decides which peer to try to connect to when there is a slot free?

**Comment by arvidn** [comment]
https://github.com/arvidn/libtorrent/blob/RC_1_2/src/peer_list.cpp#L450

**Comment by arvidn** [comment]
here's the comparison function to rank peers: https://github.com/arvidn/libtorrent/blob/RC_1_2/src/peer_list.cpp#L1275

**Comment by Sophist-UK** [comment]
Ok - I think that there is scope to improve one or both comparison functions to:

a. Include compare_peer_erase code in compare_peer - but the opposite way around. If a peer is a good erase candidate, that doesn't mean it actually gets erased, and being a good erase candidate probably makes it a bad connect candidate.

b. Add code to store download connection time and use that to calculate download speed, and use that to prioritise peers by speed - putting peers with no previous connections first so that we get speed data.

What do you think?

**Comment by Seeker2** [comment]
Side-issues your logic may run into...

> "b. Ideally I would like the BitTorrent protocol to allow seed to estimate the time before it starts downloading and to estimate the maximum upload speed, so that LibTorrent can decide whether to hold on to that connection or try to find another connection which will start faster and / or download faster (using the above algorithm for the latter"

Allow [a] seed to estimate its own max upload speed...or allow [YOUR] not-yet-a-seed to estimate its own max upload speed?

> "1.) Most people do know their download capacity"

Far fewer know their upload capacity, which can put extreme limits on their download speeds as well! (Thanks mainly to bufferbloat and Tit-For-Tat behavior.)

> "1. You can't know your own download capacity."

Even if your download speed has reached the max download speed allowed?
Or rather...when that happens, should download from slow or fast peers be decreased proportionally, based on piece rarity, based on torrent type/priority, and/or some other criteria?

> "2A) Yes - it is possible that later a connection which is currently slow may get much faster. However... A) do we want to wait on the off chance that it will speed up, or would it be better to cut our losses and find a faster peer?"

Are you getting better than you're giving? If so, even if somewhat slow (like 10-30 KB/sec)...others probably won't give as generously.
Speeds between otherwise equal peers can vary even due to pingtimes. A fast peer halfway around the world may look slow due to bufferbloat, but careful max speed limits and uTP rate behavior should help.

> "2. A slow peer now may become fast later (if demand for its bandwidth is reduced)."

On very large torrents -- typically >10 GB size...
I've seen peers that upload nothing to me for over an hour...but shortly after I start uploading to them at 20-50 KB/sec, they give me 1+ MB/sec in return back.
Best I can tell...they have few upload slots but a very high upload speed, and it was only when I was "barely faster" than the upload from their slowest upload slot peer that I got anything at all.
(BitTyrant-type logic capitalized on conditionally-fast peers, but such logic has been removed from libtorrent...and was not to my knowledge enabled by default in libtorrent even when it was present.)
So do you mark such peers as slow, fast, or unknown/untested?

Numerous peers (and seeds) have multiple ip addresses, which may clutter the ip lists that libtorrent needs to try to find other working peers/seeds.
Such peers may have both IPv4 and IPv6 addresses, LAN ips, and even link-local network shares that may "leak" into the ip lists that libtorrent uses.
Other peers may only be on a IPv4 address a few minutes before jumping to another. Their rarity is made up for by how quickly they can flood an ip list with forever-dead ips.
Many/most peers on public tracker public torrents are firewalled, so barring uTP hole-punching they are hopelessly firewalled.

On smaller-sized (<1 GB?) torrents, fast peers can join, download, and leave so quickly that they only serve to add more dead ips to the ip list. Try them quick or not at all!
Priority attempting of ips just received from a PEX message may help connect to short-lived superfast peers.

On very busy probably-larger torrents, peers and especially seeds can already have max connections and won't accept your incoming attempts even if you try often.

All-combined, I experience less than 20% success rate on the ip list with smaller/busy public torrents.

It'll be a lot quicker to search through 10000 ips if many could be regularly removed or blacklisted from outgoing attempts.
...If only there was some way for peers to report their ip changes or report if they're hopelessly firewalled so as to not have old/dead/firewalled ip addresses get into the ip list in the first place!
A second ip list may be needed of ips to NOT try!

All the effort to keep only the fastest peers connected may be wasted if incoming connections are constantly added and/or displacing them.
But if not allowing incoming connections for long periods of time, (like >15 minutes) it's safe to assume the remaining connected peers are likely not the fastest in the swarm.

> "Obviously, for torrents with a limited number of seeds, a slow download is better than none."

Typically, allowed upload speed is a tiny fraction of max download speed -- reduced even further when fast downloading eats into the upload bandwidth overhead reserved for uploading (part bufferbloat, part download given higher priority).

The transition from peer to seed typically sees every seed connection dropped and blacklisted from reconnect, followed by a rather quick fill-in of extra peers back to max connections per torrent.
Uploading to all the newly-connected peers at once might result in very low upload speeds per peer, but not uploading to many at once will mean most get nothing.
**This is default behavior for basically every BitTorrent client in existence.**

If seeds connected to and uploaded to far fewer peers at 5+ KB/sec each instead of to many at <1 KB/sec each...the end result would be faster downloads for almost everyone:
https://github.com/qbittorrent/qBittorrent/issues/2193

As a default rule, adding another connected peer to a busy torrent should not have as high a priority as a poorly-seeded one. (Otherwise, the rich get richer and the poor get poorer.)

**Comment by Sophist-UK** [comment]
@Seeker2 

> > "b. Ideally I would like the BitTorrent protocol to allow seed to estimate the time before it starts downloading and to estimate the maximum upload speed, so that LibTorrent can decide whether to hold on to that connection or try to find another connection which will start faster and / or download faster (using the above algorithm for the latter"

I apologise for not making my intent clear - I am saying it would be great if the standard bitTorrent protocol gave this information when a connection is made in order to allow the uploader to give a hint to the downloader about how long they are likely to need to wait before the send starts and the likely maximum bandwidth this will be sent at. This would allow the downloader BT client to make a judgement about whether to wait or not. However AFAIK this is NOT part of the BT protocol and is unlikely to be added.

> Or rather...when that happens, should download from slow or fast peers be decreased proportionally, based on piece rarity, based on torrent type/priority, and/or some other criteria?

I agree that when global connections are spread between different torrents, it would be nice to have some sort of algorithm that decides how to spread them between peers. Personally, in most cases (but not all) I would not want Peer_Turnover to disconnect the fastest connections on one torrent just because they are significantly slower than the slowest connections on a different torrent. I will have to consider how this can be accommodated by the above proposal for a more intelligent Peer_Turnover algorithm.

> > "2A) Yes - it is possible that later a connection which is currently slow may get much faster. However... A) do we want to wait on the off chance that it will speed up, or would it be better to cut our losses and find a faster peer?"
> 
> Are you getting better than you're giving? If so, even if somewhat slow (like 10-30 KB/sec)...others probably won't give as generously.
> Speeds between otherwise equal peers can vary even due to pingtimes. A fast peer halfway around the world may look slow due to bufferbloat, but careful max speed limits and uTP rate behavior should help.

This is NOT the way that TCP should work - it is supposed to have an ack window so that the round trip times do not affect transfer speeds - but in practice this may happen. When you talk about bufferbloat, is this something that is known to happen? I have no idea how uTP works.

> I've seen peers that upload nothing to me for over an hour...but shortly after I start uploading to them at 20-50 KB/sec, they give me 1+ MB/sec in return back.
> Best I can tell...they have few upload slots but a very high upload speed, and it was only when I was "barely faster" than the upload from their slowest upload slot peer that I got anything at all.
> (BitTyrant-type logic capitalized on conditionally-fast peers, but such logic has been removed from libtorrent...and was not to my knowledge enabled by default in libtorrent even when it was present.)
> So do you mark such peers as slow, fast, or unknown/untested?

The first question is whether you want LibTorrent to prioritise uploads to peers it is trying to download from, in order to try to improve download speeds from BT clients that prioritise downloads for those who upload. If so, then LibTorrent connections to Peers rather than Seeds should immediately try to offer an upload and see if that helps trigger a download. This is outside the scope of what I am going to be looking at.

> Numerous peers (and seeds) have multiple ip addresses, which may clutter the ip lists that libtorrent needs to try to find other working peers/seeds.
> Such peers may have both IPv4 and IPv6 addresses, LAN ips, and even link-local network shares that may "leak" into the ip lists that libtorrent uses.
> Other peers may only be on a IPv4 address a few minutes before jumping to another. Their rarity is made up for by how quickly they can flood an ip list with forever-dead ips.
> Many/most peers on public tracker public torrents are firewalled, so barring uTP hole-punching they are hopelessly firewalled.
> On smaller-sized (<1 GB?) torrents, fast peers can join, download, and leave so quickly that they only serve to add more dead ips to the ip list. Try them quick or not at all!

If you know of any other way of handling these other than trying to connect max_failcount times and then discarding the IP address then please suggest it. Perhaps the default value of 3 should be lower.

> Priority attempting of ips just received from a PEX message may help connect to short-lived superfast peers.

This might be helpful, but I am not intending to look at this.

> On very busy probably-larger torrents, peers and especially seeds can already have max connections and won't accept your incoming attempts even if you try often.

I am not sure what the BT protocol is for this. Do they accept the connection and then close it immediately or simply reject the connection in the first place? If they simply reject the connection attempt, can LibTorrent distinguish between an attempt to connect to a non-existent IP address, an existent IP address with no opening for the port, an IP address with no client listening on the port and an IP:port which exists but the client refuses the connection? And is a client which exists and connected but which is no longer sharing handled differently.

> All-combined, I experience less than 20% success rate on the ip list with smaller/busy public torrents.

The faster that LibTorrent removes these from the peer list the better. As I said, perhaps max_failcount should be 2 or even 1 rather than the default 3.
 
> It'll be a lot quicker to search through 10000 ips if many could be regularly removed or blacklisted from outgoing attempts.

LibTorrent removes IPs which have failed to connect max_failcount times. Reduce this value to make these IPs get removed faster. But beware if LibTorrent cannot distinguish between non-existent clients and clients that are too-busy to connect.

> ...If only there was some way for peers to report their ip changes or report if they're hopelessly firewalled so as to not have old/dead/firewalled ip addresses get into the ip list in the first place!

They don't. 

> A second ip list may be needed of ips to NOT try!

IMO not a good idea. Better to try quickly and move on - this sort of list will stop you connecting to a client which e.g. went away for a reboot even after they readvertise.

> All the effort to keep only the fastest peers connected may be wasted if incoming connections are constantly added and/or displacing them.

Unclear how an incoming connection displaces a connection which is kept during Peer_Turnover.

> But if not allowing incoming connections for long periods of time, (like >15 minutes) it's safe to assume the remaining connected peers are likely not the fastest in the swarm.

Unclear what you mean.

> If seeds connected to and uploaded to far fewer peers at 5+ KB/sec each instead of to many at <1 KB/sec each...the end result would be faster downloads for almost everyone:
> [qbittorrent/qBittorrent#2193](https://github.com/qbittorrent/qBittorrent/issues/2193)

Something different - not going to look at implementing this.

> As a default rule, adding another connected peer to a busy torrent should not have as high a priority as a poorly-seeded one. (Otherwise, the rich get richer and the poor get poorer.)

I agree that we should consider this when disconnecting existing connections in Peer_Turnover and when deciding what torrents to attempt new connections on.

**Comment by arvidn** [comment]
> I am saying it would be great if the standard bitTorrent protocol gave this information when a connection is made in order to allow the uploader to give a hint to the downloader about how long they are likely to need to wait before the send starts and the likely maximum bandwidth this will be sent at. This would allow the downloader BT client to make a judgement about whether to wait or not. However AFAIK this is NOT part of the BT protocol and is unlikely to be added.

Do you have an idea of how to implement this?

BitTorrent clients don't have a queue of peers where you wait for your turn to receive data. Typically the decision of which peers to unchoke is made every 15 seconds, only taking into account the number of bytes transferred in those last 15 seconds.

Both the wait time and the rate depends in impossibly many factors, not just local but in remote peers as well. I don't think there's a reliable way of estimating a transfer rate a peer can achieve before trying.

**Comment by Sophist-UK** [comment]
> Do you have an idea of how to implement this?

No. It is NOT part of the BT protocol AFAIK.

> Both the wait time and the rate depends in impossibly many factors, not just local but in remote peers as well. I don't think there's a reliable way of estimating a transfer rate a peer can achieve before trying.

I agree - and you can't even reliably predict future speed based on past speed. But that is the best we have to go on.

**Comment by arvidn** [comment]
>> Do you have an idea of how to implement this?
> 
> No. It is NOT part of the BT protocol AFAIK.

I know, I asked if you have any ideas of how to implement your suggested extension to the protocol.

**Comment by Seeker2** [comment]
> it would be great if the standard bitTorrent protocol gave this information when a connection is made in order to allow the uploader to give a hint to the downloader about how long they are likely to need to wait before the send starts and the likely maximum bandwidth this will be sent at. This would allow the downloader BT client to make a judgment about whether to wait or not. However AFAIK this is NOT part of the BT protocol and is unlikely to be added.

Upload slot logic totally gets in the way of making any estimates for how long a wait might be. If that's improved, this addition isn't needed as badly.
(Not saying you are...but...) If you implement waiting queues, you will get very long waits -- see eDonkey for an example!

> I agree that when global connections are spread between different torrents, it would be nice to have some sort of algorithm that decides how to spread them between peers. Personally, in most cases (but not all) I would not want Peer_Turnover to disconnect the fastest connections on one torrent just because they are significantly slower than the slowest connections on a different torrent. I will have to consider how this can be accommodated by the above proposal for a more intelligent Peer_Turnover algorithm.

Other BitTorrent clients (uTorrent/Tixati/Vuze?) have torrent-level priority user options to help decide how to allocate resources. These controls are often ineffective -- if a busy torrent has 100+ connected seeds/peers and a barely-active one has only 1 seed+1 peer, then setting the barely-active one to HIGHEST priority and the busy torrent to LOWEST priority doesn't change much. Setting every torrent to HIGHEST priority is also ineffective.

> When you talk about bufferbloat, is this something that is known to happen? I have no idea how uTP works.

Yes, bufferbloat often appears in shared networks. It's more noticeable when >50% load.
https://en.wikipedia.org/wiki/Bufferbloat
From:  https://www.bufferbloat.net/projects/bloat/wiki/Introduction/
"we think there is a possibility that bufferbloat ...could be a greater threat to the Internet than the congestion-driven near-collapse of the NSF backbone in 1986."
"there is a potential for nearly incalculable havoc in the worst case, but we don’t even know in principle how to estimate the overall risk."

There's also a big difference on how uTP should work and how uTP works in practice:
Poor uTP performance    https://github.com/arvidn/libtorrent/issues/3542

> The first question is whether you want LibTorrent to prioritise uploads to peers it is trying to download from, in order to try to improve download speeds from BT clients that prioritise downloads for those who upload. If so, then LibTorrent connections to Peers rather than Seeds should immediately try to offer an upload and see if that helps trigger a download. This is outside the scope of what I am going to be looking at.

Prioritizing upload to peers that upload to your peer is Tit-For-Tat logic, which is a core specification in the BitTorrent protocol.
If you get 10-30 KB/sec from a peer but are only uploading to it at half that speed, you are probably getting a good deal if that torrent is poorly seeded. Dropping that peer connection to allow in "potentially better" peers is probably a bad idea.
Due to numerous severe problems in numerous BitTorrent clients' seed logic, peers are generally better for download speed.

"On very busy probably-larger torrents, peers and especially seeds can already have max connections and won't accept your incoming attempts even if you try often."

> I am not sure what the BT protocol is for this. Do they accept the connection and then close it immediately or simply reject the connection in the first place? If they simply reject the connection attempt, can LibTorrent distinguish between an attempt to connect to a non-existent IP address, an existent IP address with no opening for the port, an IP address with no client listening on the port and an IP:port which exists but the client refuses the connection? And is a client which exists and connected but which is no longer sharing handled differently.

uTorrent's logs explicitly says "disconnected, is seed" if an incoming seed connection connects to your seed. From what I can tell, qBitTorrent (which uses libtorrent) works the same way.
Both seem to have a memory of seeds to not reconnect to. This is annoying when I'm doing repeated tests using the same torrent -- I'll complete it, stop it, delete its contents, do a force recheck, and try to download it again...but uTorrent and qBitTorrent already marked my connection as a seed and won't allow it to reconnect. I often have to completely remove the torrent to get the 2nd and subsequent tests to work!

"All-combined, I experience less than 20% success rate on the ip list with smaller/busy public torrents."

> If you know of any other way of handling these other than trying to connect max_failcount times and then discarding the IP address then please suggest it. Perhaps the default value of 3 should be lower.
> The faster that LibTorrent removes these from the peer list the better. As I said, perhaps max_failcount should be 2 or even 1 rather than the default 3.
> LibTorrent removes IPs which have failed to connect max_failcount times. Reduce this value to make these IPs get removed faster. But beware if LibTorrent cannot distinguish between non-existent clients and clients that are too-busy to connect.

Large peer/seed ip lists are going to be unwieldy no matter what's done...but that logic sounds decent.

"A second ip list may be needed of ips to NOT try!"

> IMO not a good idea. Better to try quickly and move on - this sort of list will stop you connecting to a client which e.g. went away for a reboot even after they readvertise.

Incoming connections would not be blocked by the 2nd ip list -- it's only a means to triage a long ip list by not attempting outgoing connections to "likely dead" ips. An incoming connection from an ip on that list will remove that ip from that list. PEX lists should probably remove "not dead" ips from the 2nd ip list as well.
If LibTorrent is removing failed ips from the 1st list, won't the same ips just get re-added by PEX lists?

"...If only there was some way for peers to report their ip changes or report if they're hopelessly firewalled so as to not have old/dead/firewalled ip addresses get into the ip list in the first place!"

> They don't.

The Peer EXchange (PEX) specifications explicitly mentions that only working peers+seeds should be included in the PEX ip:port lists which get passed on to other peers/seeds. Quite a few BitTorrent apps don't follow that correctly or sanity-check the results. (like 0.0.0.0 , 127.0.0.1 , LAN ips, and multicast ips get included)

> Unclear how an incoming connection displaces a connection which is kept during Peer_Turnover.

The (like >15 minutes) means I don't mean a single Peer_Turnover interval.
Over an hour, even some/many peers you'd like to stay connected to will disconnect spontaneously. I even fought with that problem in local loopback tests with multiple BitTorrent clients running on the same computer! (especially when testing uTP)
Will disconnected "fast peers" be high priority for immediate outgoing reconnect attempts?

There's little point in breaking a slow peer connection if it's just going to reconnect <1 minute later. And if it's something like BitComet, it may make 2-10 connection attempts to your ip in a second trying to reconnect.

If near max connections, LibTorrent and uTorrent quit making outgoing connections but accept incoming connections. This leaves some room for incoming connections, although it's possible on uTorrent to flood them out by making too many outgoing connection attempts at once (often TCP+uTP at the same time to the same ip) and reaching max limits anyway.

Max connection limits end up becoming roughly the current number of connections with "enough" started torrents. If a torrent has 1000+ dead/firewalled/busy ips, it will have many half-open outgoing connection attempts (or uTP equivalent) all the time.

**Comment by Sophist-UK** [comment]
> > > Do you have an idea of how to implement this?
> > 
> > 
> > No. It is NOT part of the BT protocol AFAIK.
> 
> I know, I asked if you have any ideas of how to implement your suggested extension to the protocol.

No idea - there may or may not be some sort of committee or group that discusses and agrees this protocol. But you would have to join the group and suggest it and see if people agree. Then there would need to be agreement on the exact protocol in a way that was backwards compatible with clients that haven't implemented it. Then some clients would implement it.

**Comment by Sophist-UK** [comment]
@Seeker2 
> The Peer EXchange (PEX) specifications explicitly mentions that only working peers+seeds should be included in the PEX ip:port lists which get passed on to other peers/seeds. Quite a few BitTorrent apps don't follow that correctly or sanity-check the results. (like 0.0.0.0 , 127.0.0.1 , LAN ips, and multicast ips get included)

I will check and see whether LibTorrent includes a check for non-routable IP addresses (which can of course be local): 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16, 100.64.0.0/10, 169.254.0.0/16, fd00::/8, fe80::/10

I think that these should be disallowed for any PEX / tracker / DHT response which does't itself originate from one of these addresses.

**Comment by arvidn** [comment]
> No idea - there may or may not be some sort of committee or group that discusses and agrees this protocol. But you would have to join the group and suggest it and see if people agree.

yes, that's https://github.com/bittorrent/bittorrent.org

> Then there would need to be agreement on the exact protocol in a way that was backwards compatible with clients that haven't implemented it. Then some clients would implement it.

I'm suggesting someone implements it *first*, to demonstrate that it works. But I wasn't really asking about the process of documenting and extending the protocol. But specifically how to implement it.

What would this extension *do* and *how* would it do it?

**Comment by Sophist-UK** [comment]
> What would this extension _do_ and _how_ would it do it?

@arvidn The purpose (if it is even possible) is to give hints to the downloading client about the way the upload client is working (choking rules, bandwidth allocation, typical waits before starting a download etc.) that will allow the downloading client to make better informed decisions about whether to prioritise upload on this connection or wait for a download to start or disconnect and try to find a better connection i.e. to minimise the time a client spends with connections that are not actually doing anything useful and which could be reused with a different connection that is useful.

Issues: Uploading client may not be able to work out these details, or may not wish to share them (for privacy reasons or because they don't want the client to know how to game the system), or they may simply not be beneficial.

**Comment by the8472** [comment]
I just wanted to note that we implemented something similar in biglybt and it works fairly well. I.e. optimistic disconnect based on (lack of) transferred data + remembering peer stats and taking them into account when choosing who to connect to.

**Comment by Sophist-UK** [comment]
I am still keen to try to make further improvements, but my previous PR has still not been merged, and I need some reassurance that my efforts will not be wasted.

**Comment by Sophist-UK** [comment]
I am still keen to try to make further improvements, but my previous PR has still not been merged, and I need some reassurance that my efforts will not be wasted.

**Comment by Sophist-UK** [comment]
I am still keen to try to make further improvements, but my previous PR has still not been merged, and I need some reassurance that my efforts will not be wasted.
